src/IJ_mv/IJMatrix.c

```diff
@@ -252,3 +242,259 @@ hypre_IJMatrixRead( const char     *filename,
    return hypre_error_flag;
 }
 
+/*--------------------------------------------------------------------------
+ * hypre_IJMatrixReadBinary
+ *
+ * Reads a matrix from file stored in binary format. The resulting IJMatrix
+ * is stored on host memory.
+ *--------------------------------------------------------------------------*/
+
+HYPRE_Int
+hypre_IJMatrixReadBinary( const char      *prefixname,
+                          MPI_Comm         comm,
+                          HYPRE_Int        type,
+                          HYPRE_IJMatrix  *matrix_ptr )
+{
+   HYPRE_IJMatrix  matrix;
+
+   /* Local buffers */
+   hypre_uint32   *i32buffer = NULL;
+   hypre_uint64   *i64buffer = NULL;
+   hypre_float    *f32buffer = NULL;
+   hypre_double   *f64buffer = NULL;
+
+   /* Matrix buffers */
+   HYPRE_Int       nrows;
+   HYPRE_BigInt   *rows;
+   HYPRE_BigInt   *cols;
+   HYPRE_Complex  *vals;
+
+   /* Local variables */
+   HYPRE_Int       myid;
+   char            filename[1024];
+   HYPRE_BigInt    ilower, iupper, jlower, jupper;
+   size_t          i, count;
+   hypre_uint64    header[8];
+   FILE           *fp;
+
+   /* Set filename */
+   hypre_MPI_Comm_rank(comm, &myid);
+   hypre_sprintf(filename, "%s.%05d.bin", prefixname, myid);
+
+   /* Open file */
+   if ((fp = fopen(filename, "rb")) == NULL)
+   {
+      hypre_error_w_msg(HYPRE_ERROR_GENERIC, "Could not open input file\n");
+      return hypre_error_flag;
+   }
+
+   /*---------------------------------------------
+    * Read header (64 bytes) from file
+    *---------------------------------------------*/
+
+   count = 8;
+   if (fread(header, sizeof(hypre_uint64), count, fp) != count)
+   {
+      hypre_error_w_msg(HYPRE_ERROR_GENERIC, "Could not read header entries\n");
+      return EXIT_FAILURE;
+   }
+
+   /* Exit if trying to read from big-endian machine */
+   if ((*(char*)&header[7]) == 0)
+   {
+      hypre_error_w_msg(HYPRE_ERROR_GENERIC, "Support to big-endian machines is incomplete!");
+      return hypre_error_flag;
+   }
+
+   /* Check for integer overflow */
+   if (header[2] > HYPRE_INT_MAX)
```

Comment:
I don't understand this `if` statement as written.  I get the idea of checking for potential overflow, but why doesn't this check against the size of `BigInt`?  I also don't understand what `INT_MAX` or `INT_MIN` actually mean (same for the others).  And as mentioned before, we should NOT associate specific numbers like 32 and 64 with `Int` or `BigInt`.  In these Print/ReadBinary routines, you are explicitly checking the sizes against 32 and 64, which totally makes sense and doesn't link 32 to `Int` or 64 to `BigInt`.

Reply:
Thanks @rfalgout!! Regarding the first question: I need to check against `HYPRE_Int` because the variable in question `nrows` will be used as input to `HYPRE_IJMatrixSetValues`, which expects it to be a `HYPRE_Int`. Now, there's the question: should it really be a `HYPRE_Int` there? I think a `HYPRE_BigInt` is more appropriate. But as long as we don't make this change, we'll need to do the overflow check with `INT_MAX`.

Second question: I agree that's wrong. I implemented a fix on [c35b5a0](https://github.com/hypre-space/hypre/pull/826/commits/c35b5a0aa7bca181ae2107816a0ec5fd645773db) to use the appropriate limit values according to the data types: `int` and `long long int`. Let me know if that's Ok! There's still the question about moving that block of code to some other place...

Reply:
Well, I answered my question about what `MAX` and `MIN` mean here.  I was confused because you are introducing these to avoid numbers that are too large, whereas `MIN` is a negative number.  I guess I don't really think you need to introduce these in the first place.  If you do, however, they should be in `general.h` and start with lower-case `hypre_`.

Good question about `SetValues`.  My guess is we thought this would never be larger than 2 billion since it refers to the local number of rows on the processor.  Maybe that's not a good assumption?  I think I agree with you.

But, since you are casting `nrows` to `HYPRE_Int`, why not just check against `sizeof(HYPRE_Int)`?

Reply:
Never mind my question about `sizeof`...

Reply:
Okay, sorry for all the fuss.  I think you have it right just the way it is. :)  I would only suggest moving the `HYPRE_MPI_BIG_INT` and `HYPRE_MPI_INT` defines next to each other instead of interleaving them with the other 4 new defines (put the `MPI` ones at the bottom like we do for the Real stuff).

There is still the question about the `SetValues` interface.  We can maybe take that offline with the rest of the team for now.

Reply:
Thanks for all the comments, Rob! I'll move those macros close to each other. And I agree we can talk about `SetValues` later :)